pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -0,0 +1,142 @@
+	return factory.New().ResyncEvery(time.Minute).WithInformers(
+		kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Informer(),
+		kubeInformers.InformersFor("").Core().V1().Nodes().Informer(),
+		networkInformer.Informer(),
+		operatorClient.Informer(),
+	).WithSync(c.sync).ResyncEvery(time.Minute).ToController("QuorumGuardController", eventRecorder.WithComponentSuffix("quorum-guard-controller"))
+}
+
+func (c *QuorumGuardController) isInHAMode(ctx context.Context) (bool, error) {
+	// update once
+	if c.haMode == "" {
+		infraData, err := c.infraClient.Infrastructures().Get(ctx, infrastructureClusterName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		// Added to make sure that iif ha mode is not set in infrastructure object
+		// we will get the default value configv1.FullHighAvailabilityMode
+		if infraData.Status.HighAvailabilityMode == "" {
+			klog.Infof("HA mode was not set in infrastructure resource setting it to default value %s", configv1.FullHighAvailabilityMode)
+			infraData.Status.HighAvailabilityMode = configv1.FullHighAvailabilityMode
+		}
+		c.haMode = infraData.Status.HighAvailabilityMode
+		klog.Infof("HA mode is: %s", c.haMode)
+	}
+
+	return c.haMode == configv1.FullHighAvailabilityMode, nil
+}
+
+func (c *QuorumGuardController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.ensureEtcdGuardDeployment(ctx, syncCtx.Recorder())
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "QuorumGuardControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Error",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("QuorumGuardControllerUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "QuorumGuardControllerDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *QuorumGuardController) ensureEtcdGuardDeployment(ctx context.Context, recorder events.Recorder) error {
+	haMode, err := c.isInHAMode(ctx)
+	if err != nil {
+		klog.Infof("Failed to validate ha mode %v ", err)
+		return err
+	} else if !haMode {
+		return nil
+	}
+
+	_, err = c.kubeClient.AppsV1().Deployments(operatorclient.TargetNamespace).Get(ctx,
+		EtcdGuardDeploymentName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		recorder.Eventf("NoEtcdGuardDeployment", "etcd guard deployment was not found, creating one")
+		return c.applyDeployment(recorder)
+	case err != nil:
+		klog.Warningf("failed to get %s deployment err %v", EtcdGuardDeploymentName, err.Error())
+		return err
+	default:
+		return nil
+	}
+}
+
+// Applying etcd guard deployment
+func (c *QuorumGuardController) applyDeployment(recorder events.Recorder) error {
```

Comment:
in the future (soon), this manifest will not be static. We must define the initial deployment and PDB based on replica count for the control-plane as defined by the install-config. So for example, if we supported 5 node in the future we would want the replicas of the deployment to accurately reflect.

In the future we will need to react to changes in the node count. For example, I install 5 node cluster then scale down to 3 node. We would need to observe that and update the deployment and pdb to reflect. I dont think that is in the scope of this PR but it should be considered in the design.

Reply:
Not sure i understand how it can be dynamic. It means that we need to verify number of masters every time node is added or removed.It is possible though i think it can be done with a new ticket.

Reply:
Yes in the future, out of scope of this PR